doc/source/whatsnew/v0.24.2.rst

```diff
@@ -52,7 +52,7 @@ Bug Fixes
 **I/O**
 
 - Bug in reading a HDF5 table-format ``DataFrame`` created in Python 2, in Python 3 (:issue:`24925`)
--
+- Bug in retaining timezone in :class:`DateTimeIndex` when saving and loading the :class"`DataFrame` from ``msgpack``. (:issue:`25148`)
```

Comment:
DatetimeIndex, can you use ``:meth:read_msgpack`` here

Reply:
> there must be something going else wrong here. The data is stored as utc and the tz is stored as well. what exactly is the problem?

Yes we were encoding tz using two value (zone as a str) and (Time value in utc format).
As per pandas DateTimeIndex documentation tz could be
```
tz : pytz.timezone or dateutil.tz.tzfile
```
The current encoding version is capable of encoding pytz.timezone but not the dateutil timezones.
Infact code will throw this exception when we try that.

Reply:
```python
>>> df1 = pd.DataFrame({'A':pd.date_range(start="2018-08-01",end="2018-08-30",tz=dateutil.tz.tzutc())})
>>> df1.to_msgpack("~/xyz.msg")
>>> a = pd.read_msgpack("~/xyz.msg")
Traceback (most recent call last):
  File "/home/chandan/Softwares/pandas/pandas/core/dtypes/common.py", line 2037, in pandas_dtype
    npdtype = np.dtype(dtype)
TypeError: Invalid datetime unit in metadata string "[ns, tzutc()]"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/chandan/Softwares/pandas/pandas/io/packers.py", line 207, in read_msgpack
    return read(fh)
  File "/home/chandan/Softwares/pandas/pandas/io/packers.py", line 187, in read
    unpacked_obj = list(unpack(fh, encoding=encoding, **kwargs))
  File "pandas/io/msgpack/_unpacker.pyx", line 483, in pandas.io.msgpack._unpacker.Unpacker.__next__
    return self._unpack(unpack_construct, None, 1)
  File "pandas/io/msgpack/_unpacker.pyx", line 410, in pandas.io.msgpack._unpacker.Unpacker._unpack
    ret = execute(&self.ctx, self.buf, self.buf_tail, &self.buf_head)
  File "/home/chandan/Softwares/pandas/pandas/io/packers.py", line 667, in decode
    blocks = [create_block(b) for b in obj[u'blocks']]
  File "/home/chandan/Softwares/pandas/pandas/io/packers.py", line 667, in <listcomp>
    blocks = [create_block(b) for b in obj[u'blocks']]
  File "/home/chandan/Softwares/pandas/pandas/io/packers.py", line 648, in create_block
    b[u'compress']), b[u'shape'])
  File "/home/chandan/Softwares/pandas/pandas/io/packers.py", line 322, in unconvert
    dtype = pandas_dtype(dtype).base
  File "/home/chandan/Softwares/pandas/pandas/core/dtypes/common.py", line 2043, in pandas_dtype
    dtype))
TypeError: data type 'datetime64[ns, tzutc()]' not understood
```